File: python/benchmark.py
import sys
import os
import time
import asyncio
import queue
import multiprocessing
import janus
import json
import collections

# Add the parent folder to path to find common module
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from common.data_loader import load_data
from common.stats import calculate_stats, print_stats

async def benchmark_asyncio_queue(data, total_bytes):
    q = asyncio.Queue()
    times_ns = []

    # Producer
    async def producer():
        for item in data:
            await q.put(item)

    # Consumer
    async def consumer():
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = await q.get()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            q.task_done()
            received += 1

    await asyncio.gather(producer(), consumer())
    stats = calculate_stats(times_ns, total_bytes)
    print_stats("asyncio.Queue", stats)

def benchmark_queue(data, total_bytes):
    import threading
    q = queue.Queue()
    times_ns = []

    def producer():
        for item in data:
            q.put(item)

    def consumer():
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = q.get()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            q.task_done()
            received += 1

    prod_thread = threading.Thread(target=producer)
    cons_thread = threading.Thread(target=consumer)

    prod_thread.start()
    cons_thread.start()

    prod_thread.join()
    cons_thread.join()

    stats = calculate_stats(times_ns, total_bytes)
    print_stats("queue.Queue", stats)

async def benchmark_janus(data, total_bytes):
    q = janus.Queue()
    times_ns = []

    async def producer():
        for item in data:
            await q.async_q.put(item)

    async def consumer():
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = await q.async_q.get()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            q.async_q.task_done()
            received += 1

    await asyncio.gather(producer(), consumer())
    stats = calculate_stats(times_ns, total_bytes)
    print_stats("janus", stats)

def benchmark_multiprocessing_queue(data, total_bytes):
    q = multiprocessing.Queue()
    times_ns = multiprocessing.Manager().list()

    def producer(q, data):
        for item in data:
            q.put(item)

    def consumer(q, expected, times_ns):
        received = 0
        while received < expected:
            t0 = time.perf_counter_ns()
            _ = q.get()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            received += 1

    prod_proc = multiprocessing.Process(target=producer, args=(q, data))
    cons_proc = multiprocessing.Process(target=consumer, args=(q, len(data), times_ns))

    prod_proc.start()
    cons_proc.start()

    prod_proc.join()
    cons_proc.join()

    stats = calculate_stats(list(times_ns), total_bytes)
    print_stats("multiprocessing.Queue", stats)

import redis
def benchmark_redis_py(data, total_bytes):
    try:
        r = redis.Redis(host='localhost', port=6379, db=0)
        r.ping()
    except redis.ConnectionError:
        print("Skipping redis-py benchmark: Local Redis not running")
        return

    queue_name = 'test_queue'
    r.delete(queue_name)
    times_ns = []

    def producer():
        for item in data:
            r.rpush(queue_name, item)

    def consumer():
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = r.blpop(queue_name)
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            received += 1

    import threading
    prod_thread = threading.Thread(target=producer)
    cons_thread = threading.Thread(target=consumer)

    prod_thread.start()
    cons_thread.start()

    prod_thread.join()
    cons_thread.join()

    stats = calculate_stats(times_ns, total_bytes)
    print_stats("redis-py (local Redis)", stats)

import zmq
import zmq.asyncio
async def benchmark_zmq_asyncio(data, total_bytes):
    ctx = zmq.asyncio.Context()
    times_ns = []

    async def producer():
        sock = ctx.socket(zmq.PUSH)
        sock.bind("inproc://asyncio_test")
        await asyncio.sleep(0.1)
        for item in data:
            await sock.send_string(item)
        sock.close()

    async def consumer():
        sock = ctx.socket(zmq.PULL)
        sock.connect("inproc://asyncio_test")
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = await sock.recv_string()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            received += 1
        sock.close()

    await asyncio.gather(consumer(), producer())
    ctx.term()
    stats = calculate_stats(times_ns, total_bytes)
    print_stats("zmq.asyncio (inproc)", stats)

import aiorwlock
async def benchmark_aiorwlock_deque(data, total_bytes):
    queue = collections.deque()
    lock = aiorwlock.RWLock()
    not_empty = asyncio.Event()
    times_ns = []

    async def producer():
        for item in data:
            async with lock.writer_lock:
                queue.append(item)
                not_empty.set()

    async def consumer():
        received = 0
        while received < len(data):
            await not_empty.wait()
            t0 = time.perf_counter_ns()
            async with lock.writer_lock:
                if queue:
                    _ = queue.popleft()
                    if not queue:
                        not_empty.clear()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            received += 1

    await asyncio.gather(producer(), consumer())
    stats = calculate_stats(times_ns, total_bytes)
    print_stats("aiorwlock + deque", stats)


async def benchmark_aioqueue(data, total_bytes):
    q = asyncio.Queue() # aioqueue behaves exactly as asyncio.Queue for basics
    times_ns = []

    async def producer():
        for item in data:
            await q.put(item)

    async def consumer():
        received = 0
        while received < len(data):
            t0 = time.perf_counter_ns()
            _ = await q.get()
            t1 = time.perf_counter_ns()
            times_ns.append(t1 - t0)
            q.task_done()
            received += 1

    await asyncio.gather(producer(), consumer())
    stats = calculate_stats(times_ns, total_bytes)
    print_stats("aioqueue", stats)

def main():
    data_path = sys.argv[1] if len(sys.argv) > 1 else "../datasets/test_small.jsonl"
    if not os.path.exists(data_path):
        data_path = os.path.join("datasets", "test_small.jsonl")
    print(f"Loading data from: {data_path}")
    data = load_data(data_path)
    total_bytes = sum(len(item) for item in data)
    print(f"Loaded {len(data)} records, total size: {total_bytes} bytes")

    asyncio.run(benchmark_asyncio_queue(data, total_bytes))
    benchmark_queue(data, total_bytes)
    asyncio.run(benchmark_janus(data, total_bytes))
    benchmark_multiprocessing_queue(data, total_bytes)
    benchmark_redis_py(data, total_bytes)
    asyncio.run(benchmark_zmq_asyncio(data, total_bytes))
    asyncio.run(benchmark_aiorwlock_deque(data, total_bytes))


    print("Python benchmarks completed.")

if __name__ == "__main__":
    main()

File: javascript/benchmark.js
const fs = require('fs');
const path = require('path');
const { loadData } = require('../common/data_loader');
const { calculateStats, printStats } = require('../common/stats');
const fastq = require('fastq');

async function benchmarkArrayQueue(data, totalBytes) {
    const queue = [];
    const times_ns = [];

    // Producer
    for (const item of data) {
        queue.push({ item, start: process.hrtime.bigint() });
    }

    // Consumer
    let received = 0;
    while (received < data.length) {
        const task = queue.shift();
        const end = process.hrtime.bigint();
        times_ns.push(Number(end - task.start));
        received++;
    }

    const stats = calculateStats(times_ns, totalBytes);
    printStats("Array.shift()", stats);
}

async function benchmarkFastq(data, totalBytes) {
    const times_ns = [];
    let received = 0;

    return new Promise((resolve) => {
        const q = fastq(function worker(task, cb) {
            const end = process.hrtime.bigint();
            times_ns.push(Number(end - task.start));
            received++;
            cb(null);
        }, 1);

        q.drain = function() {
            if (received === data.length) {
                const stats = calculateStats(times_ns, totalBytes);
                printStats("fastq", stats);
                resolve();
            }
        };

        for (const item of data) {
            const start = process.hrtime.bigint();
            q.push({ item, start });
        }
    });
}

async function main() {
    let dataPath = process.argv[2] || '../datasets/test_small.jsonl';
    if (!fs.existsSync(dataPath)) {
        dataPath = path.join(__dirname, '..', 'datasets', 'test_small.jsonl');
    }
    console.log(`Loading data from: ${dataPath}`);

    const data = loadData(dataPath);
    const totalBytes = data.reduce((acc, val) => acc + Buffer.byteLength(val, 'utf8'), 0);
    console.log(`Loaded ${data.length} records, total size: ${totalBytes} bytes`);


    await benchmarkArrayQueue(data, totalBytes);
    await benchmarkFastq(data, totalBytes);
    await benchmarkPQueue(data, totalBytes);
    await benchmarkNpmQueue(data, totalBytes);
    await benchmarkRxjs(data, totalBytes);
    // Skipping bullmq, bee-queue, better-queue, workerpool, Node.js MessageChannel as they require
    // external services like Redis or process forks which are beyond simple micro-benchmark scope here.

    console.log("JavaScript benchmarks completed.");
}


main().catch(console.error);

const { default: PQueue } = require('p-queue');
async function benchmarkPQueue(data, totalBytes) {
    const queue = new PQueue({ concurrency: 1 });
    const times_ns = [];

    for (const item of data) {
        const start = process.hrtime.bigint();
        queue.add(async () => {
            const end = process.hrtime.bigint();
            times_ns.push(Number(end - start));
        });
    }

    await queue.onIdle();
    const stats = calculateStats(times_ns, totalBytes);
    printStats("p-queue", stats);
}



const { default: Queue } = require('queue');
async function benchmarkNpmQueue(data, totalBytes) {
    const q = new Queue({ concurrency: 1, autostart: true });
    const times_ns = [];

    return new Promise((resolve) => {
        let received = 0;

        q.addEventListener('end', () => {
            const stats = calculateStats(times_ns, totalBytes);
            printStats("queue (npm)", stats);
            resolve();
        });

        for (const item of data) {
            const start = process.hrtime.bigint();
            q.push((cb) => {
                const end = process.hrtime.bigint();
                times_ns.push(Number(end - start));
                received++;
                if (cb) cb();
            });
        }

        // if autostart doesn't emit end when empty
        if(data.length === 0) resolve();
    });
}

const { Subject, concatMap, of } = require('rxjs');
async function benchmarkRxjs(data, totalBytes) {
    const subject = new Subject();
    const times_ns = [];

    return new Promise((resolve) => {
        let received = 0;
        const sub = subject.pipe(
            concatMap((task) => of(task))
        ).subscribe((task) => {
            const end = process.hrtime.bigint();
            times_ns.push(Number(end - task.start));
            received++;
            if (received === data.length) {
                const stats = calculateStats(times_ns, totalBytes);
                printStats("rxjs Subject + concatMap", stats);
                sub.unsubscribe();
                resolve();
            }
        });

        for (const item of data) {
            subject.next({ item, start: process.hrtime.bigint() });
        }
    });
}
